### PR TITLE
feat: supporting peer exchange with nwaku

### PR DESCRIPTION
**Pull request by gabrielmer**
Fixing and testing Peer Exchange support using nwaku. 

The test `TestPeerExchange` should work once the following nwaku PRs are merged:
- https://github.com/waku-org/nwaku/pull/3141
- https://github.com/waku-org/nwaku/pull/3140
- https://github.com/waku-org/nwaku/pull/3127

Once they are merged, I will update the nwaku submodule in this PR and it will be ready to be merged

Important changes:
- [x] updated nwaku submodule
- [x] added integration with new `waku_get_my_peerid` and `waku_get_peerids_from_peerstore` libwaku's functions
- [x] expanding `WakuConfig` with more nwaku configurations
- [x] added functionalities and bug fixes related to getting peers' data
- [x] implementing `TestPeerExchange`

Once the nwaku submodule is updated and libwaku is recompiled, you can run the test by executing
```
go test -tags="use_nwaku,gowaku_no_rln" -run TestPeerExchange ./wakuv2/... -count 1 -v -json | jq -r '.Output'
```

Closes https://github.com/waku-org/nwaku/issues/3115

### CODE REVIEW

File: wakuv2/nwaku.go [hunk 1]
```diff
@@ -361,19 +369,23 @@ type WakuPubsubTopic = string
...
-	DnsDiscoveryUrl      string   `json:"dnsDiscoveryUrl,omitempty"`
-	MaxMessageSize       string   `json:"maxMessageSize,omitempty"`
+	Host        		 string `json:"host,omitempty"`
+	Port        		 int    `json:"port,omitempty"`
+	NodeKey     		 string `json:"key,omitempty"`
+	EnableRelay  		 bool   `json:"relay"`
+	LogLevel     		 string `json:"logLevel"`
+	DnsDiscovery 		 bool `json:"dnsDiscovery,omitempty"`
+	DnsDiscoveryUrl 	 string `json:"dnsDiscoveryUrl,omitempty"`
+	MaxMessageSize		 string `json:"maxMessageSize,omitempty"`
```
Comment (gabrielmer) [inline wakuv2/nwaku.go]: Not sure why this happens, in my IDE looks aligned, is there a linting tool I should use?

**Review by Ivansete-status — APPROVED** [review]
Wonderful! Thanks for it! :raised_hands:

File: wakuv2/nwaku.go (hunk 1 above)
Comment (richard-ramos) [reply wakuv2/nwaku.go]: In vscode, when installing the go extension, it asks to install gofmt then the linting happens automagically

File: wakuv2/nwaku.go [hunk 2]
```diff
@@ -486,9 +498,11 @@ func New(nodeKey *ecdsa.PrivateKey, fleet string, cfg *WakuConfig, logger *zap.L
 		fmt.Println("Error happened:", err.Error())
 	}
 
-	err = node.WakuRelaySubscribe(defaultPubsubTopic)
-	if err != nil {
-		fmt.Println("Error happened:", err.Error())
+	if cfg.EnableRelay {
+		err = node.WakuRelaySubscribe(defaultPubsubTopic)
+		if err != nil {
+			fmt.Println("Error happened:", err.Error())
```
Comment (richard-ramos) [inline wakuv2/nwaku.go]:
```suggestion
			return nil, err
```

File: wakuv2/nwaku.go [hunk 3]
```diff
@@ -2184,10 +2198,24 @@ func (w *Waku) Clean() error {
...
+	defer C.freeResp(resp)
+	C.cGoWakuGetMyPeerId(w.wakuCtx, resp)
+
+	if C.getRet(resp) == C.RET_OK {
+		
+		peerIdStr := C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))		
+		id, err := peer.Decode(peerIdStr)
+		if err != nil {
+			errMsg := "WakuGetMyPeerId - error decoding peerId: " + err.Error()
+			return "", errors.New(errMsg)
```
Comment (richard-ramos) [inline wakuv2/nwaku.go]:
```suggestion
			errMsg := "WakuGetMyPeerId - decoding peerId: %w"
			return "",  fmt.Errorf(errMsg, err)
```

File: wakuv2/nwaku.go [hunk 4]
```diff
@@ -2184,10 +2198,24 @@ func (w *Waku) Clean() error {
...
+		}
+		return id, nil
+	}
+	errMsg := "error WakuGetMyPeerId: " +
+		C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
+	return "", errors.New(errMsg)
```
Comment (richard-ramos) [inline wakuv2/nwaku.go]:
```suggestion
	errMsg := C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp))
	return "", fmt.Errorf("WakuGetMyPeerId: %s",  errMsg)
```

File: wakuv2/nwaku.go [hunk 5]
```diff
@@ -2796,7 +2856,7 @@ func (self *Waku) GetPeerIdsByProtocol(protocol string) (peer.IDSlice, error) {
 		for _, p := range itemsPeerIds {
 			id, err := peer.Decode(p)
 			if err != nil {
-				errMsg := "GetPeerIdsByProtocol - error converting string to int: " + err.Error()
+				errMsg := "GetPeerIdsByProtocol - error decoding peerId: " + err.Error()
 				return nil, errors.New(errMsg)
```
Comment (richard-ramos) [inline wakuv2/nwaku.go]:
```suggestion
				return nil, fmt.Errorf("GetPeerIdsByProtocol - decoding peerId: %w", err)
```

File: wakuv2/nwaku.go [hunk 6]
```diff
@@ -2628,18 +2656,22 @@ func (self *Waku) wakuStoreQuery(
...
+		numRecvPeers, err := strconv.ParseUint(numRecvPeersStr, 10, 64)
+		if err != nil {
+			return 0, err
+		}
+		return numRecvPeers, nil
 	}
 	errMsg := "error WakuPeerExchangeRequest: " +
 		C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-	return "", errors.New(errMsg)
+	return 0, errors.New(errMsg)
```
Comment (richard-ramos) [inline wakuv2/nwaku.go]:
```go
errMsg := C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp))
return 0, fmt.Errorf("WakuPeerExchangeRequest: %w", errMsg)
```

File: wakuv2/nwaku.go [hunk 7]
```diff
@@ -2776,6 +2806,36 @@ func (self *Waku) GetNumConnectedPeers(paramPubsubTopic ...string) (int, error)
...
+				return nil, errors.New(errMsg)
+			}
+			peers = append(peers, id)
+		}
+
+		return peers, nil
+	}
+	errMsg := "error GetPeerIdsFromPeerStore: " +
+		C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
+	return nil, errors.New(errMsg)
```
Comment (richard-ramos) [inline wakuv2/nwaku.go]:
```suggestion
	errMsg := C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp))
	return nil, fmt.Errorf("GetPeerIdsFromPeerStore: %w", errMsg)
```

File: wakuv2/nwaku.go [hunk 8]
```diff
@@ -2776,6 +2806,36 @@ func (self *Waku) GetNumConnectedPeers(paramPubsubTopic ...string) (int, error)
...
+		}
+		// peersStr contains a comma-separated list of peer ids
+		itemsPeerIds := strings.Split(peersStr, ",")
+
+		var peers peer.IDSlice
+		for _, peerId := range itemsPeerIds {
+			id, err := peer.Decode(peerId)
+			if err != nil {
+				errMsg := "GetPeerIdsFromPeerStore - error decoding peerId: " + err.Error()
+				return nil, errors.New(errMsg)
```
Comment (richard-ramos) [inline wakuv2/nwaku.go]:
```suggestion
				return nil, fmt.Errorf("GetPeerIdsFromPeerStore - decoding peerId: %w", err)
```

**Review by richard-ramos — APPROVED** [review]

**Pull request merged**